### PR TITLE
[Geometry Checker] Add method to check if geometry has curves

### PR DESCRIPTION
**Pull request by lbartoletti**
## Description

In GIS, even when the layer type supports curves, it's relevant to check geometries that might contain them.
This verification is crucial to ensure certain operations.

In this algorithm following the logic of others, we propose 2 resolution methods alongside the conventional NoAction approach.:
- deleting the feature
- converting curves into straight curves

Funded by QGIS (Grant OpenSource 2023) and Oslandia

Part of https://github.com/qgis/QGIS-Enhancement-Proposals/issues/236

### CODE REVIEW

File: src/analysis/vector/geometry_checker/qgsgeometrycurvecheck.cpp [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+  Q_UNUSED( messages )
+  const QMap<QString, QgsFeatureIds> featureIds = ids.isEmpty() ? allLayerFeatureIds( featurePools ) : ids.toMap();
+  const QgsGeometryCheckerUtils::LayerFeatures layerFeatures( featurePools, featureIds, compatibleGeometryTypes(), feedback, mContext );
+  for ( const QgsGeometryCheckerUtils::LayerFeature &layerFeature : layerFeatures )
+  {
+    const QgsAbstractGeometry *geom = layerFeature.geometry().constGet();
+    for ( int iPart = 0, nParts = geom->partCount(); iPart < nParts; ++iPart )
+    {
+      const QgsAbstractGeometry *part = QgsGeometryCheckerUtils::getGeomPart( geom, iPart );
+      if ( part->hasCurvedSegments() )
```
Comment (benoitdm-oslandia) [inline src/analysis/vector/geometry_checker/qgsgeometrycurvecheck.cpp:32]: Are you collecting errors or curved segments?

File: src/analysis/vector/geometry_checker/qgsgeometrycurvecheck.cpp [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+      {
+        errors.append( new QgsGeometryCheckError( this, layerFeature, part->centroid(), QgsVertexId( iPart ) ) );
+      }
+    }
+  }
+}
+
+void QgsGeometryCurveCheck::fixError( const QMap<QString, QgsFeaturePool *> &featurePools, QgsGeometryCheckError *error, int method, const QMap<QString, int> & /*mergeAttributeIndices*/, Changes &changes ) const
```
Comment (benoitdm-oslandia) [inline src/analysis/vector/geometry_checker/qgsgeometrycurvecheck.cpp:40]: `int method` is an enum, why are you using an `int`?

File: src/analysis/vector/geometry_checker/qgsgeometrycurvecheck.h [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    Q_DECLARE_TR_FUNCTIONS( QgsGeometryCurveCheck )
+  public:
+
+    /**
+     * Creates a new has curve check with the provided \a context. No options are supported in \a configuration.
+     */
+    explicit QgsGeometryCurveCheck( QgsGeometryCheckContext *context, const QVariantMap &configuration )
+      : QgsGeometryCheck( context, configuration ) {}
+    QList<Qgis::GeometryType> compatibleGeometryTypes() const override { return factoryCompatibleGeometryTypes(); }
+    void collectErrors( const QMap<QString, QgsFeaturePool *> &featurePools, QList<QgsGeometryCheckError *> &errors, QStringList &messages, QgsFeedback *feedback, const LayerFeatureIds &ids = LayerFeatureIds() ) const override;
```
Comment (benoitdm-oslandia) [inline src/analysis/vector/geometry_checker/qgsgeometrycurvecheck.h:40]: I just understand `collectErrors` is overridden from `QgsGeometryCheck`, but can you add a comment to explain you will collect curved geometry in order to fix them and not collecting errors :) ?

File: src/analysis/vector/geometry_checker/qgsgeometrycurvecheck.h [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    void fixError( const QMap<QString, QgsFeaturePool *> &featurePools, QgsGeometryCheckError *error, int method, const QMap<QString, int> &mergeAttributeIndices, Changes &changes ) const override;
```
Comment (benoitdm-oslandia) [inline src/analysis/vector/geometry_checker/qgsgeometrycurvecheck.h:41]: The same, just to explain why kind of fixes are availables (what they do)

File: tests/src/geometry_checker/testqgsgeometrychecks.cpp [hunk 5]
```diff
@@ -329,6 +332,60 @@ void TestQgsGeometryChecks::testContainedCheck()
...
+  QVERIFY( fixCheckError( testContext.second,  errs2[0],
+                          QgsGeometryCurveCheck::StraightCurves, QgsGeometryCheckError::StatusFixed,
+  {{errs2[0]->layerId(), errs2[0]->featureId(), QgsGeometryCheck::ChangeFeature, QgsGeometryCheck::ChangeChanged, QgsVertexId()}} ) );
+  QgsFeature f2;
+  const bool valid2 = testContext.second[errs2[0]->layerId()]->getFeature( errs2[0]->featureId(), f2 );
+  QVERIFY( valid2 );
+  QVERIFY( ! f2.geometry().constGet()->hasCurvedSegments() );
+
+  cleanupTestContext( testContext );
+}
```
Comment (benoitdm-oslandia) [inline tests/src/geometry_checker/testqgsgeometrychecks.cpp:387]: Can you add another test to check the "do-nothing" case?

File: src/analysis/vector/geometry_checker/qgsgeometrycurvecheck.cpp (hunk 2 above)
Comment (lbartoletti) [reply src/analysis/vector/geometry_checker/qgsgeometrycurvecheck.cpp:40]: Old code. Herited from https://github.com/qgis/QGIS/blob/b2685bf0e96e9367c2e486488ad9319a3324848d/src/analysis/vector/geometry_checker/qgsgeometrycheck.h#L285

File: src/analysis/vector/geometry_checker/qgsgeometrycurvecheck.cpp (hunk 1 above)
Comment (lbartoletti) [reply src/analysis/vector/geometry_checker/qgsgeometrycurvecheck.cpp:32]: In this case, curved segments "is" the error.
As other Checker, I collect the centroid of the geometry

File: src/analysis/vector/geometry_checker/qgsgeometrycurvecheck.h (hunk 3 above)
Comment (lbartoletti) [reply src/analysis/vector/geometry_checker/qgsgeometrycurvecheck.h:40]: > curved geometry in order to fix them and not collecting errors

It's the same, here :)

File: src/analysis/vector/geometry_checker/qgsgeometrycurvecheck.h (hunk 4 above)
Comment (lbartoletti) [reply src/analysis/vector/geometry_checker/qgsgeometrycurvecheck.h:41]: They are mentioned in ResolutionMethod

File: tests/src/geometry_checker/testqgsgeometrychecks.cpp (hunk 5 above)
Comment (lbartoletti) [reply tests/src/geometry_checker/testqgsgeometrychecks.cpp:394]: Done in [0a9c00b](https://github.com/qgis/QGIS/pull/55590/commits/0a9c00b0cebd5e29ef0322f903a01d07b1d85a7a)

**Pull request closed without merging**